.github/workflows/linux.yml

```diff
@@ -3,116 +3,136 @@ name: Linux build and test
 on: [push, pull_request]
 
 jobs:
-  test-clang-based:
-    name: clang ${{ matrix.clang }}, ${{ matrix.os }}, CUDA ${{matrix.cuda}}, ROCm ${{matrix.rocm}}
-    runs-on: ${{ matrix.os }}
+  prep-container:
+    name: Prepare CI container clang ${{ matrix.clang }}, ${{ matrix.os }}, CUDA ${{matrix.cuda_major_version}}.${{matrix.cuda_minor_version}}.${{matrix.cuda_patch_version}}, ROCm ${{matrix.rocm}}
+    runs-on: ubuntu-22.04
+    permissions:
+      pull-requests: read
+      packages: write
+      actions: read
+    outputs:
+      docker_matrix: ${{ steps.matrix-output.outputs.json }}
     strategy:
       matrix:
-        clang: [15, 16, 17, 18]
+        clang: [15, 16, 17, 18, 19, 20]
         os: [ubuntu-22.04]
-        cuda: [11.0.2]
+        cuda_major_version: [11]
+        cuda_minor_version: [0]
+        cuda_patch_version: [2]
         rocm: [5.4.3]
         include:
           - clang: 17
             os: ubuntu-22.04
-            cuda: 11.0.2
+            cuda_major_version: 11
+            cuda_minor_version: 0
+            cuda_patch_version: 2
             rocm: 5.6.1
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          submodules: 'recursive'
+          fetch-depth: 20
+      - uses: ./.github/workflows/create-ci-docker
+        id: docker
+        with:
+          LLVM_VERSION: ${{ matrix.clang }}
+          ROCM_VERSION: ${{ matrix.rocm }}
+          CUDA_MAJOR_VERSION: ${{ matrix.cuda_major_version }}
+          CUDA_MINOR_VERSION: ${{ matrix.cuda_minor_version }}
+          CUDA_PATCH_VERSION: ${{ matrix.cuda_patch_version }}
+          BASE_TAG: ${{ matrix.os }}
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+      - uses: beacon-biosignals/matrix-output@v1
+        id: matrix-output
+        with:
+          yaml: |
+            name: ${{ matrix.clang }}-${{ matrix.os }}
+            clang: ${{ matrix.clang }}
+            os: ${{ matrix.os }}
+            cuda: ${{matrix.cuda_major_version}}.${{matrix.cuda_minor_version}}.${{matrix.cuda_patch_version}}
+            rocm: ${{matrix.rocm}}
+            image: ${{ steps.docker.outputs.docker_image }}
+
+  test-clang-based:
+    name: clang ${{ matrix.docker.clang }}, ${{ matrix.test_kind }}, CUDA ${{matrix.docker.cuda}}, ROCm ${{matrix.docker.rocm}}, ${{ matrix.docker.os }}
+    runs-on: ${{ matrix.docker.os }}
+    needs: prep-container
+    strategy:
+      matrix:
+        docker: ${{fromJson(needs.prep-container.outputs.docker_matrix)}}
+        test_kind: [cpu, sscp, gpu]
```

Comment:
Perhaps we should consider renaming this, as SSCP is also CPU/GPU in a way.... perhaps `omp, sscp, smcp`? Or `omp, sscp, smcp-interop`?
 But we can also leave it as is for the time being, not a strong issue for me :)